examples/src/java/com/twitter/heron/examples/api/StatefulSlidingWindowTopology.java

```diff
@@ -0,0 +1,171 @@
+//  Copyright 2017 Twitter. All rights reserved.
+//
+//  Licensed under the Apache License, Version 2.0 (the "License");
+//  you may not use this file except in compliance with the License.
+//  You may obtain a copy of the License at
+//
+//  http://www.apache.org/licenses/LICENSE-2.0
+//
+//  Unless required by applicable law or agreed to in writing, software
+//  distributed under the License is distributed on an "AS IS" BASIS,
+//  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+//  See the License for the specific language governing permissions and
+//  limitations under the License.
+
+package com.twitter.heron.examples.api;
+
+import java.util.Map;
+import java.util.Random;
+import java.util.logging.Logger;
+
+import com.twitter.heron.api.Config;
+import com.twitter.heron.api.HeronSubmitter;
+import com.twitter.heron.api.bolt.BaseStatefulWindowedBolt;
+import com.twitter.heron.api.bolt.BaseWindowedBolt;
+import com.twitter.heron.api.bolt.OutputCollector;
+import com.twitter.heron.api.spout.BaseRichSpout;
+import com.twitter.heron.api.spout.SpoutOutputCollector;
+import com.twitter.heron.api.state.State;
+import com.twitter.heron.api.topology.IStatefulComponent;
+import com.twitter.heron.api.topology.OutputFieldsDeclarer;
+import com.twitter.heron.api.topology.TopologyBuilder;
+import com.twitter.heron.api.topology.TopologyContext;
+import com.twitter.heron.api.tuple.Fields;
+import com.twitter.heron.api.tuple.Tuple;
+import com.twitter.heron.api.tuple.Values;
+import com.twitter.heron.api.utils.Utils;
+import com.twitter.heron.api.windowing.TupleWindow;
+import com.twitter.heron.common.basics.ByteAmount;
+import com.twitter.heron.examples.api.bolt.PrinterBolt;
+
+/**
+ * A sample topology that demonstrates the usage of {@link com.twitter.heron.api.bolt.IStatefulWindowedBolt}
+ * to calculate sliding window sum.  Topology also demonstrates how stateful window processing is done
+ * in conjunction with effectively once guarantees
+ */
+public final class StatefulSlidingWindowTopology {
+
+  private static final Logger LOG = Logger.getLogger(StatefulSlidingWindowTopology.class.getName());
+
+  private StatefulSlidingWindowTopology() {
+  }
+
+  private static class WindowSumBolt extends BaseStatefulWindowedBolt<String, Long> {
+    private static final long serialVersionUID = -539382497249834244L;
+    private State<String, Long> state;
+    private long sum;
+
+    private OutputCollector collector;
+
+    @Override
+    public void prepare(Map<String, Object> topoConf, TopologyContext context, OutputCollector collector) {
+      this.collector = collector;
+    }
+
+    @Override
+    public void initState(State<String, Long> state) {
+      this.state = state;
+      sum = state.getOrDefault("sum", 0L);
+    }
+
+    @Override
+    public void execute(TupleWindow inputWindow) {
+      for (Tuple tuple : inputWindow.get()) {
+        sum += tuple.getLongByField("value");
+      }
+      state.put("sum", sum);
```

Comment:
You could move this line to preSave. That way the topoliogy can be used both in effectively once as well as atleast/atmost once mode

Reply:
+1